java/ql/src/experimental/Security/CWE/CWE-094/JexlInjectionLib.qll

```diff
@@ -0,0 +1,288 @@
+    m instanceof GetterMethod and
+    ma.getQualifier() = fromNode.asExpr() and
+    ma = toNode.asExpr()
+  )
+}
+
+/**
+ * Method in the `JexlEngine` class that get or set a property with a Jexl expression.
+ */
+private class JexlEngineGetSetPropertyMethod extends Method {
+  JexlEngineGetSetPropertyMethod() {
+    getDeclaringType() instanceof JexlEngine and
+    hasName(["getProperty", "setProperty"])
+  }
+}
+
+/**
+ * Defines methods that triggers direct evaluation of Jexl expressions.
+ */
+abstract private class DirectJexlEvaluationMethod extends Method { }
+
+/**
+ * A method in the `JexlExpression` class that evaluates a Jexl expression.
+ */
+private class JexlExpressionEvaluateMethod extends DirectJexlEvaluationMethod {
+  JexlExpressionEvaluateMethod() {
+    getDeclaringType() instanceof JexlExpression and hasName("evaluate")
+  }
+}
+
+/**
+ * A method in the `JexlScript` class that executes a Jexl script.
+ */
+private class JexlScriptExecuteMethod extends DirectJexlEvaluationMethod {
+  JexlScriptExecuteMethod() { getDeclaringType() instanceof JexlScript and hasName("execute") }
+}
+
+/**
+ * A method in the `JxltEngine.Expression` class that evaluates an expression.
+ */
+private class JxltEngineExpressionEvaluateMethod extends DirectJexlEvaluationMethod {
+  JxltEngineExpressionEvaluateMethod() {
+    getDeclaringType() instanceof JxltEngineExpression and hasName("evaluate")
+  }
+}
+
+/**
+ * A method in the `JxltEngine.Expression` class that evaluates the immediate sub-expressions.
+ */
+private class JxltEngineExpressionPrepareMethod extends DirectJexlEvaluationMethod {
+  JxltEngineExpressionPrepareMethod() {
+    getDeclaringType() instanceof JxltEngineExpression and hasName("prepare")
+  }
+}
+
+/**
+ * A method in the `JxltEngine.Template` class that evaluates a template.
+ */
+private class JxltEngineTemplateEvaluateMethod extends DirectJexlEvaluationMethod {
+  JxltEngineTemplateEvaluateMethod() {
+    getDeclaringType() instanceof JxltEngineTemplate and hasName("evaluate")
+  }
+}
+
+/**
+ * A method in the `UnifiedJEXL.Expression` class that evaluates a template.
+ */
+private class UnifiedJexlExpressionEvaluateMethod extends DirectJexlEvaluationMethod {
+  UnifiedJexlExpressionEvaluateMethod() {
+    getDeclaringType() instanceof UnifiedJexlExpression and hasName("evaluate")
+  }
+}
+
+/**
+ * A method in the `UnifiedJEXL.Expression` class that evaluates the immediate sub-expressions.
+ */
+private class UnifiedJexlExpressionPrepareMethod extends DirectJexlEvaluationMethod {
+  UnifiedJexlExpressionPrepareMethod() {
+    getDeclaringType() instanceof UnifiedJexlExpression and hasName("prepare")
+  }
+}
+
+/**
+ * A method in the `UnifiedJEXL.Template` class that evaluates a template.
+ */
+private class UnifiedJexlTemplateEvaluateMethod extends DirectJexlEvaluationMethod {
+  UnifiedJexlTemplateEvaluateMethod() {
+    getDeclaringType() instanceof UnifiedJexlTemplate and hasName("evaluate")
+  }
+}
```

Comment:
```suggestion
/**
 * Defines methods that triggers direct evaluation of Jexl expressions.
 */
private class DirectJexlEvaluationMethod extends Method {
  DirectJexlEvaluationMethod() {
    getDeclaringType() instanceof JexlExpression and hasName("evaluate") or
    getDeclaringType() instanceof JexlScript and hasName("execute") or
    getDeclaringType() instanceof JxltEngineExpression and hasName(["evaluate", "prepare"]) or
    getDeclaringType() instanceof JxltEngineTemplate and hasName("evaluate") or
    getDeclaringType() instanceof UnifiedJexlExpression and hasName(["evaluate", "prepare"]) or
    getDeclaringType() instanceof UnifiedJexlTemplate and hasName("evaluate")
  }
}
```

Haven't checked, but if the other types lack a `prepare` method altogether then this could becomes just a list of 6 types `and hasName(["evaluate", "prepare"])`

Reply:
Thanks for the suggestion - that makes it much shorter! Technically, I can write the following even if the types don't have all three methods:

```
private class DirectJexlEvaluationMethod extends Method {
  DirectJexlEvaluationMethod() {
    (
      getDeclaringType() instanceof JexlExpression
      or
      getDeclaringType() instanceof JexlScript
      or
      getDeclaringType() instanceof JxltEngineExpression
      or
      getDeclaringType() instanceof JxltEngineTemplate
      or
      getDeclaringType() instanceof UnifiedJexlExpression
      or
      getDeclaringType() instanceof UnifiedJexlTemplate
    ) and
    hasName(["evaluate", "execute", "prepare"])
  }
}
```

That would work. However, that looks a bit confusing to me since none of the listed types has all three methods. I'd prefer to use your version that describes the methods of the specific classes. Let me know if I am missing something.